### ISSUE TITLE
Figure3

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/total-testosterone#5]
# ============================================================
# Figure 3 maker (Colab-ready)
# - Subjects: age & BMI complete => N=5854
# - Mean imputation -> standardization -> PCA(2D for visualization)
# - KMeans(K=4) in 19D standardized space
# - Save fig as PNG/PDF + CSV outputs + ZIP download
# ============================================================

import os
import numpy as np
import pandas as pd
import zipfile

import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans

# Colab download support
try:
    from google.colab import files
    IN_COLAB = True
except Exception:
    IN_COLAB = False

# -------------------------
# 0) paths
# -------------------------
XLSX_PATH = "/content/drive/MyDrive/CommmedpatientTT.xlsx"
assert os.path.exists(XLSX_PATH), f"File not found: {XLSX_PATH}"

OUTDIR = "/content/figure3_outputs"
os.makedirs(OUTDIR, exist_ok=True)

# -------------------------
# 1) load + column harmonization
# -------------------------
df = pd.read_excel(XLSX_PATH)

target_cols = [
    "age","height","BW","BMI","Hb","Hct","Alb","AST","ALT","ALP","Cre",
    "Glu","TG","HDL-C","LDL-C","LH","FSH","CRP","TT"
]

rename_map = {
    "HDL": "HDL-C",
    "LDL": "LDL-C",
    "HDL_C": "HDL-C",
    "LDL_C": "LDL-C",
    "BodyWeight": "BW",
    "Weight": "BW",
}

df = df.rename(columns={k: v for k, v in rename_map.items() if k in df.columns})
missing_cols = [c for c in target_cols if c not in df.columns]
assert len(missing_cols) == 0, f"Missing required columns: {missing_cols}"

X_raw = df[target_cols].copy()

print("Loaded:", df.shape)
print("Columns:", df.columns.tolist())
print("\nX_raw:", X_raw.shape)
print("NA counts (top 10):")
print(X_raw.isna().sum().sort_values(ascending=False).head(10))

# -------------------------
# 2) (optional) BMI reconstruction from height/BW
#    - fills only when height and BW exist
# -------------------------
print("\n[A] BMI reconstruction from height/BW (fills only if possible)")
h = pd.to_numeric(X_raw["height"], errors="coerce")
w = pd.to_numeric(X_raw["BW"], errors="coerce")

height_is_cm = h.median(skipna=True) > 3.0
h_m = h / 100.0 if height_is_cm else h

bmi_from_hw = w / (h_m ** 2)

n_bmi_missing_before = int(X_raw["BMI"].isna().sum())
mask_bmi_na = X_raw["BMI"].isna()
X_raw.loc[mask_bmi_na, "BMI"] = bmi_from_hw.loc[mask_bmi_na]
n_bmi_missing_after = int(X_raw["BMI"].isna().sum())

print("height unit assumed:", "cm" if height_is_cm else "m")
print("BMI missing before:", n_bmi_missing_before)
print("BMI missing after :", n_bmi_missing_after)

# -------------------------
# 3) Subjects filter: age & BMI complete (N=5854 expected)
# -------------------------
print("\n[Subjects filter]")
n_before = len(X_raw)
n_missing_age = int(X_raw["age"].isna().sum())
n_missing_bmi = int(X_raw["BMI"].isna().sum())
print("Missing age:", n_missing_age)
print("Missing BMI:", n_missing_bmi)

mask_complete_age_bmi = X_raw["age"].notna() & X_raw["BMI"].notna()
X_subj = X_raw.loc[mask_complete_age_bmi].copy()

print("Rows before:", n_before)
print("Rows after age/BMI filter:", len(X_subj))
N_SUBJECTS = len(X_subj)

# -------------------------
# 4) Impute -> Standardize
# -------------------------
imputer = SimpleImputer(strategy="mean")
scaler = StandardScaler()

X_imp = imputer.fit_transform(X_subj.values)
X_std = scaler.fit_transform(X_imp)
print("\nX_std shape:", X_std.shape)

# -------------------------
# 5) KMeans in 19D (cluster assignment)
#    - If you want the exact "校正4" behavior, keep random_state fixed.
#    - If you prefer seed scan to minimize tiny clusters, set USE_SEED_SCAN=True.
# -------------------------
USE_SEED_SCAN = True
K = 4
N_INIT = 50
MIN_CLUSTER_SIZE = 30  # try to avoid very small clusters; may still fail due to data structure

def run_kmeans(seed: int):
    km = KMeans(n_clusters=K, n_init=N_INIT, random_state=seed)
    labels = km.fit_predict(X_std)
    unique, counts = np.unique(labels, return_counts=True)
    sizes = {int(u): int(c) for u, c in zip(unique, counts)}
    min_size = min(sizes.values())
    return labels, sizes, min_size

if not USE_SEED_SCAN:
    seed_chosen = 0
    labels, sizes, min_size = run_kmeans(seed_chosen)
else:
    print("\n[KMeans seed scan]")
    seeds = list(range(0, 51))  # 0..50
    best = None  # (min_size, seed, labels, sizes)
    for s in seeds:
        lab, sz, ms = run_kmeans(s)
        # Prefer solutions meeting MIN_CLUSTER_SIZE; within that, maximize min cluster size
        ok = (ms >= MIN_CLUSTER_SIZE)
        score = (1 if ok else 0, ms)  # lexicographic
        if best is None or score > best[0]:
            best = (score, s, lab, sz)

    (score, seed_chosen, labels, sizes) = best
    if score[0] == 0:
        print("[WARN] No solution met MIN_CLUSTER_SIZE. Falling back to best available.")
    print("[Chosen KMeans solution]")
    print("Chosen seed:", seed_chosen)
    print("Cluster sizes:", sizes)
    print("MIN_CLUSTER_SIZE:", MIN_CLUSTER_SIZE)

# Save cluster sizes
size_path = os.path.join(OUTDIR, "cluster_sizes.csv")
pd.DataFrame({"cluster": list(sizes.keys()), "n": list(sizes.values())}).to_csv(size_path, index=False)

# -------------------------
# 6) PCA for visualization (2D only)
# -------------------------
pca2 = PCA(n_components=2, random_state=0)
X_pca2 = pca2.fit_transform(X_std)

evr = pca2.explained_variance_ratio_
pc1 = evr[0] * 100
pc2 = evr[1] * 100
pc12 = (evr[0] + evr[1]) * 100

print("\n[PCA explained variance (2D)]")
print(f"PC1 (%): {pc1:.2f}")
print(f"PC2 (%): {pc2:.2f}")
print(f"PC1+PC2 (%): {pc12:.2f}")

pca_path = os.path.join(OUTDIR, "pca_explained_variance.csv")
pd.DataFrame({
    "component": ["PC1", "PC2"],
    "explained_variance_ratio": evr,
    "explained_variance_percent": evr * 100,
    "cumulative_percent": np.cumsum(evr) * 100
}).to_csv(pca_path, index=False)

# -------------------------
# 7) Make Figure 3
#    - PC2 axis fixed to [-5, 20]
#    - Singleton cluster (if any) highlighted with 'x' marker
# -------------------------
fig, ax = plt.subplots(figsize=(7.2, 5.4), dpi=200)

labels_arr = np.asarray(labels)

# identify singleton clusters
unique, counts = np.unique(labels_arr, return_counts=True)
singleton_clusters = set(unique[counts == 1].tolist())

for c in sorted(unique.tolist()):
    mask = (labels_arr == c)
    n_c = int(mask.sum())

    # singleton style
    if c in singleton_clusters:
        ax.scatter(X_pca2[mask, 0], X_pca2[mask, 1],
                   marker="x", s=60, linewidths=2,
                   label=f"Cluster {c} (singleton, n=1)")
    else:
        ax.scatter(X_pca2[mask, 0], X_pca2[mask, 1],
                   s=10, alpha=0.7,
                   label=f"Cluster {c} (n={n_c})")

ax.set_xlabel(f"PC1 ({pc1:.2f}%)")
ax.set_ylabel(f"PC2 ({pc2:.2f}%)")
ax.set_ylim(-5, 20)  # as per manuscript
ax.set_title("Figure 3. PCA visualization of K-means clusters (K=4)")

ax.legend(loc="best", fontsize=8, frameon=True)
ax.grid(True, linewidth=0.3, alpha=0.5)

fig_path_png = os.path.join(OUTDIR, "Figure3_PCA_KMeans.png")
fig_path_pdf = os.path.join(OUTDIR, "Figure3_PCA_KMeans.pdf")
plt.tight_layout()
plt.savefig(fig_path_png, bbox_inches="tight")
plt.savefig(fig_path_pdf, bbox_inches="tight")
plt.show()

print("\nSaved Figure 3:")
print(fig_path_png)
print(fig_path_pdf)

# -------------------------
# 8) Save meta + ZIP
# -------------------------
meta_path = os.path.join(OUTDIR, "analysis_meta.csv")
pd.DataFrame({
    "N_before": [n_before],
    "N_after_ageBMI_filter": [N_SUBJECTS],
    "missing_age_before": [n_missing_age],
    "missing_BMI_before": [n_bmi_missing_before],
    "missing_BMI_after_reconstruct": [n_bmi_missing_after],
    "height_unit_assumed": ["cm" if height_is_cm else "m"],
    "kmeans_seed_chosen": [seed_chosen],
    "kmeans_n_init": [N_INIT],
    "kmeans_K": [K],
    "PC1_percent": [round(pc1, 2)],
    "PC2_percent": [round(pc2, 2)],
    "PC1_PC2_cum_percent": [round(pc12, 2)],
    "singleton_clusters_present": [",".join(map(str, sorted(list(singleton_clusters)))) if len(singleton_clusters) else ""],
}).to_csv(meta_path, index=False)

zip_path = os.path.join(OUTDIR, "figure3_outputs.zip")
with zipfile.ZipFile(zip_path, "w", compression=zipfile.ZIP_DEFLATED) as zf:
    for fname in [
        "Figure3_PCA_KMeans.png",
        "Figure3_PCA_KMeans.pdf",
        "cluster_sizes.csv",
        "pca_explained_variance.csv",
        "analysis_meta.csv",
    ]:
        zf.write(os.path.join(OUTDIR, fname), arcname=fname)

print("\nZIP created:", zip_path)

if IN_COLAB:
    files.download(zip_path)
else:
    print("Not in Colab: download manually from:", zip_path)
